experimental/uefi/channel/src/lib.rs

```diff
@@ -66,16 +118,205 @@ impl Frame {
     }
 }
 
-pub struct Framed<T: Read + Write> {
+/// A [`Message`] is conceptual unit of data that is sent over the communication
+/// channel. It can represent either a full request, or full response of an oak_idl
+/// invocation. Requests and responses can be converted to a [`Message`], and
+/// vice versa. Usually a single [`Message`] is sent over the communication
+/// channel over several [`Frame`]'s.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Message {
+    pub method_or_status: MethodOrStatus,
+    pub body: Vec<u8>,
+}
+
+/// Deconstruct a [`Message`] into a set of [`Frame`]s.
+impl<'a> From<Message> for Vec<Frame> {
+    fn from(message: Message) -> Self {
+        let frame_bodies: Vec<Vec<u8>> = match message.body.len() {
+            // Messages with an empty body are handled differently, since
+            // chunking an empty [`Vec`] would result in 0 chunks.
+            0 => vec![vec![]],
+            _ => message
+                .body
+                .chunks(MAX_FRAME_BODY_SIZE)
+                // TODO(#2848): It'd be nice if we didn't have to reallocate here.
+                // We may be able to avoid doing so by making [`Frame`] use
+                // slices and lifetimes. Or alternatively use the Bytes crate for
+                // reference counting.
+                .map(|frame_body| frame_body.to_vec())
+                .collect(),
+        };
+        let length = frame_bodies.len();
+        frame_bodies
+            .into_iter()
+            .enumerate()
+            .map(|(index, body)| {
+                let flag = {
+                    if index == 0 {
+                        match length {
+                            1 => Flag::AtomicFrame,
+                            _ => Flag::StreamStart,
+                        }
+                    } else if index == length - 1 {
+                        Flag::StreamEnd
+                    } else {
+                        Flag::StreamContinuance
+                    }
+                };
+                Frame {
+                    method_or_status: message.method_or_status,
+                    flag,
+                    body,
+                }
+            })
+            .collect()
+    }
+}
+
+/// Struct to incrementally build a [`Message`] from a set of [`Frame`]s.
+#[derive(Default, Debug, PartialEq, Eq)]
+struct PartialMessage {
+    inner: Option<Message>,
+}
+#[derive(Debug, PartialEq, Eq)]
+enum MessageReconstructionErrors {
+    ExpectedStartFrame,
+    DoubleStartFrame,
+    MismatchingFrameHeader,
+}
+impl MessageReconstructionErrors {
+    fn as_str(&self) -> &'static str {
+        match self {
+            MessageReconstructionErrors::ExpectedStartFrame => "Invalid frame order. The first frame must be a start frame.",
+            MessageReconstructionErrors::DoubleStartFrame => "Invalid frame order. The frame method/status header does must be the same for all frames belonging to the same message.",
```

Comment:
nit: "... header must be ..."

Also, the description for this seems to be swapped with the next one.

Reply:
thanks for catching :)